### PR TITLE
RavenDB-13338: Add SkipAddingIdFieldToDynamicObject to DocumentConven…

### PR DESCRIPTION
**Pull request by garayx**
…tions

### CODE REVIEW

File: src/Raven.Client/Documents/Conventions/DocumentConventions.cs [hunk 1]
```diff
@@ -246,6 +249,21 @@ public Action<JsonSerializer> CustomizeJsonSerializer
             }
         }
 
+        /// <summary>
+        ///     If 'true' Id field will not be added to dynamic objects.
+        ///     If 'false' Id field will be added to dynamic objects.
```
Comment (ayende) [inline src/Raven.Client/Documents/Conventions/DocumentConventions.cs]: Can remove this line

File: src/Raven.Client/Documents/Conventions/DocumentConventions.cs [hunk 2]
```diff
@@ -246,6 +249,21 @@ public Action<JsonSerializer> CustomizeJsonSerializer
             }
         }
 
+        /// <summary>
+        ///     If 'true' Id field will not be added to dynamic objects.
```
Comment (ayende) [inline src/Raven.Client/Documents/Conventions/DocumentConventions.cs]:
```suggestion
        ///     By default, the field 'Id' field will be added to dynamic objects, this allows to disable this behavior.
```

That said, let's avoid the double negative here.
Make the convention: "AddIdFieldToDynamicObjects" and default to true

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Conventions/DocumentConventions.cs (hunk 2 above)
Comment (garayx) [reply src/Raven.Client/Documents/Conventions/DocumentConventions.cs]: done

File: src/Raven.Client/Documents/Conventions/DocumentConventions.cs (hunk 1 above)
Comment (garayx) [reply src/Raven.Client/Documents/Conventions/DocumentConventions.cs]: done

**Review by ayende — APPROVED** [review]

File: src/Raven.Client/Documents/Session/EntityToBlittable.cs [hunk 3]
```diff
@@ -283,7 +285,26 @@ private static bool TryRemoveIdentityProperty(BlittableJsonReaderObject document
...
+                {
+                    try
+                    {
+                        if (document.Modifications == null)
+                            document.Modifications = new DynamicJsonValue(document);
+
+                        document.Modifications.Remove("Id");
+                        return true;
+                    }
+                    catch (RuntimeBinderException)
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/EntityToBlittable.cs]: No need to catch RuntimeBinderException. We are not performing any operations on 'dynamic'

File: src/Raven.Client/Documents/Session/EntityToBlittable.cs [hunk 4]
```diff
@@ -283,7 +285,26 @@ private static bool TryRemoveIdentityProperty(BlittableJsonReaderObject document
         {
             var identityProperty = conventions.GetIdentityProperty(entityType);
             if (identityProperty == null)
+            {
+                if (conventions.AddIdFieldToDynamicObjects && entityType.GetInterfaces().Contains(typeof(IDynamicMetaObjectProvider)))
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Session/EntityToBlittable.cs]: We need to cache this, it is too expensive. Probably better to pass the entity from the calling method and just to `entity is IDynamicMetaObjectProvider`

**Review by ppekrol — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Session/EntityToBlittable.cs (hunk 4 above)
Comment (ayende) [reply src/Raven.Client/Documents/Session/EntityToBlittable.cs]: Yes, passing it there is much better

Comment (garayx) [reply src/Raven.Client/Documents/Session/EntityToBlittable.cs]: done

File: src/Raven.Client/Documents/Session/EntityToBlittable.cs (hunk 3 above)
Comment (garayx) [reply src/Raven.Client/Documents/Session/EntityToBlittable.cs]: done

File: test/FastTests/Issues/RavenDB_13338.cs [hunk 5]
```diff
@@ -40,5 +40,53 @@ public void SkipAddingIdFieldToDynamicObjectShouldWork()
...
+                    Assert.Equal("1", loadJObject.User.ToString());
+                    Assert.Equal(jObjectId, loadJObject.Id.ToString());
+                }
+
+                using (var commands = store.Commands())
+                {
+                    const string idField = "Id";
+
+                    var bJsonExpandoObject = commands.Get(expandoObjectId);
+                    Assert.Equal(false, bJsonExpandoObject.Contains(idField));
```
Comment (ppekrol) [inline test/FastTests/Issues/RavenDB_13338.cs]: 1. Use `ContainsKey` method
2. `Assert.False(bJsonExpandoObject.ContainsKey(idField))`

Comment (garayx) [reply test/FastTests/Issues/RavenDB_13338.cs]: done

**Review by ppekrol — APPROVED** [review]

**Pull request merged**